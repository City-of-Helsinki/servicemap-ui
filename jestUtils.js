import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { ThemeProvider } from '@mui/material/styles';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import themes from './src/themes';
import english from './src/i18n/en';

const mockStore = configureStore([]);

// Mock props for intl provider
const intlMock = {
  locale: 'en',
  messages: english,
  wrapRichTextChunksInFragment: false,
};

const Providers = (mockState) => ({ children }) => {
  const store = mockStore(mockState);
  return (
    <Provider store={store}>
      <IntlProvider {...intlMock}>
        <ThemeProvider theme={themes.SMTheme}>
          {children}
        </ThemeProvider>
      </IntlProvider>
    </Provider>
  );
}

export const getRenderWithProviders = mockState => component => render(component, { wrapper: Providers(mockState) });
