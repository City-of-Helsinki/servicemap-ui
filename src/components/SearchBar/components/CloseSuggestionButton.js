import React from 'react';
import PropTypes from 'prop-types';
import { Typography } from '@mui/material';
import { visuallyHidden } from '@mui/utils';
import { FormattedMessage } from 'react-intl';

// Close suggestions button for suggestions dropdowns
export const CloseSuggestionButton = ({
  onClick,
  onKeyPress,
  onKeyDown,
  icon,
  srOnly,
  ...rest
}) => {
  const tabIndex = srOnly ? '-1' : '0';

  try {
    if (srOnly) {
      if (!onKeyPress) {
        throw Error('Missing onKeyPress prop for CloseSuggestionButton');
      }
      return (
        <Typography
          role="button"
          tabIndex={tabIndex}
          onClick={onClick}
          onKeyPress={onKeyPress}
          style={visuallyHidden}
          {...rest}
        >
          <FormattedMessage id="search.suggestions.hideButton" />
        </Typography>
      );
    }
    if (!onKeyDown) {
      throw Error('Missing onKeyDown prop for CloseSuggestionButton');
    }
    return (
      <div
        role="button"
        tabIndex={tabIndex}
        onClick={onClick}
        onKeyDown={onKeyDown}
        {...rest}
      >
        <Typography variant="body2"><FormattedMessage id="search.suggestions.hideButton" /></Typography>
        {!!icon ? icon: null}
      </div>
    );
  } catch (e) {
    console.error(e);
  }
  return null;
}

CloseSuggestionButton.propTypes =Â {
  onClick: PropTypes.func.isRequired,
  onKeyPress: PropTypes.func,
  onKeyDown: PropTypes.func,
  icon: PropTypes.node,
  srOnly: PropTypes.bool,
};
