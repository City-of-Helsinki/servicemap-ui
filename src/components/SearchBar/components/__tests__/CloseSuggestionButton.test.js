// CloseSuggestionButton.test.js
import React from 'react';
import { fireEvent } from '@testing-library/react';
import { ArrowDownward } from '@mui/icons-material';
import { CloseSuggestionButton } from '../CloseSuggestionButton';
import { getRenderWithProviders } from '../../../../../jestUtils';
import englishTranslations from '../../../../i18n/en';

// Generic required props for ResultItem
const mockProps = {
  onClick: () => {},
  onKeyPress: () => {},
  onKeyDown: () => {},
  icon: <ArrowDownward />,
  srOnly: false,
};

const renderWithProviders = getRenderWithProviders({});

describe('<CloseSuggestionButton />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<CloseSuggestionButton {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('simulates click event', () => {
    const mockCallBack = jest.fn();
    const { getByRole } = renderWithProviders(
      <CloseSuggestionButton {...mockProps} onClick={mockCallBack} />
    );

    fireEvent.click(getByRole('button'));
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });

  it('simulates keyboard event', () => {
    const mockCallBack = jest.fn();
    const { getAllByRole } = renderWithProviders(
      <>
        <CloseSuggestionButton {...mockProps} onKeyDown={mockCallBack} />
        <CloseSuggestionButton {...mockProps} onKeyPress={mockCallBack} srOnly />
      </>,
    );

    fireEvent.keyDown(getAllByRole('button')[0], {
      key: 'Enter',
      code: 'Enter',
      keyCode: 13,
      charCode: 13,
    });
    fireEvent.keyDown(getAllByRole('button')[0], {
      key: 'Space',
      code: 'Space',
      keyCode: 32,
      charCode: 32,
    });

    expect(mockCallBack.mock.calls.length).toEqual(2);

    // Sr only element
    fireEvent.keyPress(getAllByRole('button', { selector: 'p' })[1], {
      key: 'Enter',
      code: 'Enter',
      keyCode: 13,
      charCode: 13,
    });
    fireEvent.keyPress(getAllByRole('button', { selector: 'p' })[1], {
      key: 'Space',
      code: 'Space',
      keyCode: 32,
      charCode: 32,
    });
    expect(mockCallBack.mock.calls.length).toEqual(4);
  });

  it('does show text correctly', () => {
    const { container } = renderWithProviders(<CloseSuggestionButton {...mockProps} />);

    expect(container.querySelector('p')).toHaveTextContent(`${englishTranslations['search.suggestions.hideButton']}`);
  });

  it('does use accessibility attributes correctly', () => {
    // Visible element
    const { getAllByRole } = renderWithProviders(
      <>
        <CloseSuggestionButton {...mockProps} />
        <CloseSuggestionButton {...mockProps} srOnly />
      </>,
    );

    expect(getAllByRole('button')[0]).toHaveAttribute('role', 'button');
    expect(getAllByRole('button')[0]).not.toHaveAttribute('aria-hidden');
    expect(getAllByRole('button')[0]).toHaveAttribute('tabindex', '0');

    // SrOnly element
    expect(getAllByRole('button')[1]).toHaveAttribute('role', 'button');
    expect(getAllByRole('button')[1]).not.toHaveAttribute('aria-hidden');
    expect(getAllByRole('button')[1]).toHaveAttribute('tabindex', '-1');
  });
});
