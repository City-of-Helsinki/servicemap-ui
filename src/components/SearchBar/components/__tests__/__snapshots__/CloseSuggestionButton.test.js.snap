// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CloseSuggestionButton /> should work 1`] = `
<div>
  <div
    role="button"
    tabindex="0"
  >
    <p
      class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
    >
      Hide the list of suggestions
    </p>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-6yqi1c-MuiSvgIcon-root"
      data-testid="ArrowDownwardIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
      />
    </svg>
  </div>
</div>
`;
